src/main/java/org/truffleruby/core/string/StringNodes.java

```diff
@@ -3499,44 +3488,29 @@ public Object stringCharacterIndex(DynamicObject string, DynamicObject pattern,
                 }
             }
 
-            for(s = p, ss = pp; p < e; s = p += c, ++index) {
+            for (; p < l; p += c, ++index) {
                 c = StringSupport.preciseLength(enc, stringBytes, p, e);
                 if ( !StringSupport.MBCLEN_CHARFOUND_P(c)) return nil();
-
-                if (stringBytes[p] != patternBytes[pp]) continue;
-
-                while (p < e && pp < pe) {
-                    boolean breakOut = false;
-
-                    for (int pc = p + c; p < e && p < pc && pp < pe; ) {
-                        if (stringBytes[p] == patternBytes[pp]) {
-                            ++p;
-                            ++pp;
-                        } else {
-                            breakOut = true;
-                            break;
-                        }
-                    }
-
-                    if (breakOut) {
-                        break;
-                    }
-
-                    c = StringSupport.preciseLength(enc, stringBytes, p, e);
-                    if (! StringSupport.MBCLEN_CHARFOUND_P(c)) break;
-                }
-
-                if (pp < pe) {
-                    p = s;
-                    pp = ss;
-                } else {
+                if (matchesAtOffset(p, e, pp, pe, stringBytes, patternBytes)) {
                     return index;
                 }
             }
 
             return nil();
         }
 
+        private static boolean matchesAtOffset(int offset, int limit, int patternOffset, int patternLimit, byte[] stringBytes, byte[] patternBytes) {
```

Comment:
Other than having different limits, this looks quite similar to `ArrayUtils.memcmp`. Maybe we just need a new form of that call?